pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -448,14 +494,39 @@ private void commitSegmentMetadataInternal(String realtimeTableName,
     // Refresh the Broker routing to reflect the changes in the segment ZK metadata
     _helixResourceManager.sendSegmentRefreshMessage(realtimeTableName, committingSegmentName, false, true);
 
-    // Step-2
+    // Get current partition groups - this gives current state of latest segments for each partition
+    // E.g. [A - DONE], [B - IN_PROGRESS], [C - IN_PROGRESS]
+    PartitionLevelStreamConfig streamConfig = new PartitionLevelStreamConfig(tableConfig.getTableName(),
+        IngestionConfigUtils.getStreamConfigMap(tableConfig));
+    List<PartitionGroupMetadata> currentPartitionGroupMetadataList =
+        getCurrentPartitionGroupMetadataList(idealState, streamConfig);
+
+    // Fetches new partition groups, given current partition groups metadata.
+    // Assume stream has partitions A, B, C, all still consuming. Result will be A, B, C
+    // Assume A was split into D, E, but messages of A are yet to be consumed, result will be A, B, C
+    // Assume A was split into D, E and all messages of A are consumed, result will be B, C, D, E.
+    List<PartitionGroupInfo> newPartitionGroupInfoList =
+        getPartitionGroupInfoList(streamConfig, currentPartitionGroupMetadataList);
+    Set<Integer> newPartitionGroupSet =
+        newPartitionGroupInfoList.stream().map(PartitionGroupInfo::getPartitionGroupId).collect(Collectors.toSet());
+    int numPartitions = newPartitionGroupInfoList.size();
+
+    // Only if committingSegment's partitionGroup is present in the newPartitionGroupInfoList, we create new segment metadata
```

Comment:
There can be race conditions here? A partition may have disappeared (and so does not show up in the current list of partitions available for consumption).
I still think it is best that the segment completion logic just creates the next segment for each partition that is there in the ideal state _plus_ any new ones that it sees -- with one exception as below.
If a server gets "end-of-partition" message, it completes the segment with the end criteria set as 'end of partiition". In such end-criteria, the controller should not create a new segment for that partition.

This will be an intuitive design, and easier to reason about, since the server is the only one trying to "detect" partition ends. 

i am afraid there can be race conditions where the controller does not see the new partitions, but the server has been restarted, so it has not finished consuming the partition as yet.

Reply:
CurrentPartitionGroupMetadata is calculated using the latest segment zk metadata, for every partition group that is found in the ideal state. 

So let's say we had only partitionGroup 0, the 
`currentPartitionGroupMetadata=[(partitionGroupId:0, seqId:0, startOffset:0, endOffset: -, status:IN_PROGRESS)]. 
`

Now say this partition group 0 was committing, the currentPartitionGroupMetadata is calculated after the DONE is updated, hence 
`currentPartitionGroupMetadata=[(partitionGroupId:0, seqId:0, startOffset:0, endOffset: 130, status:DONE)].`

Now, let's say the partition group 0 started consuming for the next segment, 
`currentPartitionGroupMetadata=[(partitionGroupId:0, seqId:1, startOffset:130, endOffset: -, status:IN_PROGRESS)].`

Now, say that the partition 0 was split by the operator into 1,2. As a result, our partition group 0 detects that the partition has no more messages, and commits.
`currentPartitionGroupMetadata=[(partitionGroupId:0, seqId:1, startOffset:130, endOffset: 210, status:DONE)].
`

Next time the validation manager runs, new partition groups 1 and 2 are detected, and consuming segments are created for them.

Henceforth, the currentPartitionGroupMetadata call will always return partition group 0.
`currentPartitionGroupMetadata=[(partitionGroupId:0, seqId:1, startOffset:130, endOffset: 210, status:DONE), (partitionGroupId:1, seqId:0, startOffset:0, endOffset: -, status:IN_PROGRESS), (partitionGroupId:2, seqId:0, startOffset:0, endOffset: -, status:IN_PROGRESS)].`

So, this `A partition may have disappeared (and so does not show up in the current list of partitions available for consumption)` will never happen.

Reply:
The server is still the only one trying to detect end of partition. We are setting stopReason as "endOfPartitionGroup". Only difference is that we are relying on the controller's "new partition group info" to decide which segments to create.
The controller is not trying to detect end of partition groups explicitly. It is simply returning the new groups that it thinks should be. And we create new consuming segments for everything that is in the new partition groups list, and is not already present. It prolly looks a lil odd right now, because we only create new metadata for the committing segment. But the eventual goal is to create new metadata for all newly detected partitions.